### PR TITLE
feat(query): add arithmetic function v2 support

### LINKED ISSUE #7091: Migrate arithmetics functions to new expression framework

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

add arithmetic function v2 support.

TODO in next pr: modulo op

Summary about this PR

Fixes  #7091

### CODE REVIEW

File: common/functions-v2/src/scalars/arithmetic.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ pub fn register(registry: &mut FunctionRegistry) {
                         {
                             type T = <(L, R) as ResultTypeOfBinary>::Minus;
                             registry.register_2_arg::<NumberType<L>, NumberType<R>, NumberType<T>, _, _>(
-                                "+",
+                                "-",
```
Comment (Xuanwo) [inline common/functions-v2/src/scalars/arithmetic.rs:70]: :rofl:

File: common/codegen/src/writes/arithmetics_type_v2.rs [hunk 2]
```diff
@@ -55,16 +55,36 @@ pub fn codegen_arithmetic_type_v2() {
...
     type IntDiv: Number;
     type Modulo: Number;
     type LeastSuper: Number;
 }}
 
-pub trait ResultTypeOfUnary {{
+pub trait ResultTypeOfUnary: Sized {{
     type Negate: Number;
+    
+    fn d_checked_add(self, _rhs: Self) -> Option<Self> {{
```
Comment (andylokandy) [inline common/codegen/src/writes/arithmetics_type_v2.rs:69]: what does the `d_` mean?

File: common/codegen/src/writes/arithmetics_type_v2.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+// This code is generated by common/codegen. DO NOT EDIT.
+use super::number::Number;
+
+pub trait ResultTypeOfBinary: Sized {{
+    type AddMul: Number;
+    type Minus: Number;
+    type IntDiv: Number;
+    type Modulo: Number;
+    type LeastSuper: Number;
```
Comment (andylokandy) [inline common/codegen/src/writes/arithmetics_type_v2.rs:63]: When will it be used?

File: common/functions-v2/tests/it/scalars/testdata/arithmetic.txt [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+
+ast            : c + d
+raw expr       : +(ColumnRef(3)::UInt32, ColumnRef(4)::Float64)
+checked expr   : +<UInt32, Float64>(ColumnRef(3), ColumnRef(4))
+evaluation:
++--------+---------+------------------+---------+-----------+----------------+---------+
+|        | a       | a2               | b       | c         | d              | Output  |
++--------+---------+------------------+---------+-----------+----------------+---------+
+| Type   | Int8    | UInt8 NULL       | Int16   | UInt32    | Float64        | Float64 |
+| Domain | {1..=3} | {1..=3} ∪ {NULL} | {2..=6} | {10..=30} | {-20.0..=30.0} | Unknown |
```
Comment (andylokandy) [inline common/functions-v2/tests/it/scalars/testdata/arithmetic.txt:114]: The domain should be able to be calculated.

File: common/functions-v2/tests/it/scalars/testdata/arithmetic.txt [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+
+ast            : c * d
+raw expr       : *(ColumnRef(3)::UInt32, ColumnRef(4)::Float64)
+checked expr   : *<UInt32, Float64>(ColumnRef(3), ColumnRef(4))
+evaluation:
++--------+---------+------------------+---------+-----------+----------------+---------+
+|        | a       | a2               | b       | c         | d              | Output  |
++--------+---------+------------------+---------+-----------+----------------+---------+
+| Type   | Int8    | UInt8 NULL       | Int16   | UInt32    | Float64        | Float64 |
+| Domain | {1..=3} | {1..=3} ∪ {NULL} | {2..=6} | {10..=30} | {-20.0..=30.0} | Unknown |
```
Comment (andylokandy) [inline common/functions-v2/tests/it/scalars/testdata/arithmetic.txt:376]: ditto

File: common/functions-v2/src/scalars/arithmetic.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+use common_expression::types::number::*;
+use common_expression::types::DataType;
+use common_expression::vectorize_with_writer_2_arg;
+use common_expression::with_number_mapped_type;
+use common_expression::FunctionProperty;
+use common_expression::FunctionRegistry;
+use common_expression::NumberDomain;
+
+pub fn register(registry: &mut FunctionRegistry) {
+    registry.register_aliases("+", &["plus", "add"]);
```
Comment (andylokandy) [inline common/functions-v2/src/scalars/arithmetic.rs:26]: I think the parser won't accept `+(1, 2)` now, do we need to support it?

File: common/codegen/src/writes/arithmetics_type_v2.rs (hunk 2 above)
Comment (sundy-li) [reply common/codegen/src/writes/arithmetics_type_v2.rs]: It's just to be different with the built-in function `checked_add` in std lib.

File: common/functions-v2/src/scalars/arithmetic.rs (hunk 6 above)
Comment (sundy-li) [reply common/functions-v2/src/scalars/arithmetic.rs]: No, I will remove this op.

File: common/codegen/src/writes/arithmetics_type_v2.rs (hunk 2 above)
Comment (andylokandy) [reply common/codegen/src/writes/arithmetics_type_v2.rs]: It's weird. Can we simply name it `checked_add` just like num-traits https://docs.rs/num-traits/0.2.15/num_traits/ops/checked/trait.CheckedAdd.html

**Review by PsiACE — APPROVED** [review]

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**